File: src/sdks/core/test/suite/lifecycle.test.ts
/*
 * Copyright 2021 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the 'License');
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an 'AS IS' BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { transport } from '../../../../../test/TransportHarness.js'
import MockTransport from '../../build/javascript/src/Transport/MockTransport.mjs'
import { jest, test, expect, beforeAll } from "@jest/globals";

import { Lifecycle } from "../../build/javascript/src/firebolt";

let readyResolved: boolean = false;
let readyCalled: boolean = false;
let readyMetricCalled: boolean = false;
let readyMetricCalledAfterResolve: boolean = false;

transport.onSend((json) => {

  if (json && json.method) {

    let [module, method] = json.method.split('.');

    if (module === "Lifecycle" && method === "ready") {
      readyCalled = true;
    }
    else if (module === "Metrics" && method === "ready") {
      readyMetricCalled = true;
      if (readyResolved) {
        readyMetricCalledAfterResolve = true;
      }

    }
  }
})

const callback = jest.fn(
  (eventName) => {
    console.log(`Lifecycle Event: ${eventName}`);
  }
);
const startupState: Lifecycle.LifecycleState = Lifecycle.state();

beforeAll(() => {

  Lifecycle.once("onForeground", () => {
    Lifecycle.close(Lifecycle.CloseReason.USER_EXIT);
  });

  MockTransport.event("Lifecycle", "onInactive", null);

  Lifecycle.ready().then((_) => {
    readyResolved = true;
  }).catch((err) => {
    readyResolved = true;
    console.error("Error in Lifecycle.ready: ", err);
  });

  MockTransport.event("Lifecycle", "onForeground", null);

  let p = new Promise<void>((resolve, reject) => {
    Lifecycle.once("onUnloading", (_) => {
      setTimeout(() => {
        resolve();
      }, 100);
    });
  });
  MockTransport.event("Lifecycle", "onUnloading", null);

  return p
});

test("Lifecycle.ready Promise resolved", () => {
  expect(readyCalled).toBe(true);
  expect(readyResolved).toBe(true);
});

test("Lifecycle.ready calls Metrics.ready", () => {
  expect(readyMetricCalled).toBe(true);
  expect(readyMetricCalledAfterResolve).toBe(true);
});

test('App starts up in the "initializing" state', () => {
  expect(startupState).toBe("initializing");
});
/*
test('App moves to the "inactive" state next', () => {
  expect(callback).nthCalledWith(1, "onInactive");
});

test('App moves to the "foreground" state next', () => {
  expect(callback).nthCalledWith(2, "onForeground");
});

test('App moves to the "unloading" state next', () => {
  expect(callback).nthCalledWith(3, "onUnloading");
});
*/
test("listen() background event.", () => {
  return Lifecycle.listen("onBackground", callback).then((res: number) => {
    expect(res > 0).toBe(true);
  });
});

test("once() background event.", () => {
  return Lifecycle.once("onBackground", callback).then((res: number) => {
    expect(res > 0).toBe(true);
  });
});

test("listen() foreground event.", () => {
  return Lifecycle.listen("onForeground", callback).then((res: number) => {
    expect(res > 0).toBe(true);
  });
});

test("once() foreground event.", () => {
  return Lifecycle.once("onForeground", callback).then((res: number) => {
    expect(res > 0).toBe(true);
  });
});

test("listen() inactive event.", () => {
  return Lifecycle.listen("onInactive", callback).then((res: number) => {
    expect(res > 0).toBe(true);
  });
});

test("once() inactive event.", () => {
  return Lifecycle.once("onInactive", callback).then((res: number) => {
    expect(res > 0).toBe(true);
  });
});

test("listen() suspended event.", () => {
  return Lifecycle.listen("onSuspended", callback).then((res: number) => {
    expect(res > 0).toBe(true);
  });
});

test("once() suspended event.", () => {
  return Lifecycle.once("onSuspended", callback).then((res: number) => {
    expect(res > 0).toBe(true);
  });
});

test("listen() unloading event.", () => {
  return Lifecycle.listen("onUnloading", callback).then((res: number) => {
    expect(res > 0).toBe(true);
  });
});

test("once() unloading event.", () => {
  return Lifecycle.once("onUnloading", callback).then((res: number) => {
    expect(res > 0).toBe(true);
  });
});

test("clear()", () => {
  const result: boolean = Lifecycle.clear(-1000);
  expect(result).toBeFalsy();
});

File: src/cpp/test/unit/lifecycleTest.cpp
/*
 * If not stated otherwise in this file or this component's LICENSE file the
 * following copyright and licenses apply:
 *
 * Copyright 2025 Sky UK
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "lifecycle_impl.h"
#include "mock_helper.h"

using ::testing::_;
using ::testing::Invoke;
using ::testing::Return;

class LifecycleTest : public ::testing::Test, protected MockBase
{
protected:
    void SetUp() override {}

    void TearDown() override
    {
        EXPECT_CALL(mockHelper, unsubscribeAll(_))
            .WillRepeatedly(Invoke([](auto) { return Firebolt::Result<void>(Firebolt::Error::None); }));
    }

    Firebolt::Lifecycle::LifecycleImpl lifecycleImpl_{mockHelper};
};

TEST_F(LifecycleTest, close)
{
    nlohmann::json expectedParams;
    expectedParams["type"] = "deactivate";
    EXPECT_CALL(mockHelper, invoke("Lifecycle2.close", expectedParams))
        .WillOnce(Invoke([&](const std::string& /*methodName*/, const nlohmann::json& /*parameters*/)
                         { return Firebolt::Result<void>{Firebolt::Error::None}; }));

    auto result = lifecycleImpl_.close(Firebolt::Lifecycle::CloseType::DEACTIVATE);
    ASSERT_TRUE(result) << "Error on invoke";
}

TEST_F(LifecycleTest, state)
{
    mock_with_response("Lifecycle2.state", "initializing");
    Firebolt::Result<Firebolt::Lifecycle::LifecycleState> result = lifecycleImpl_.state();
    ASSERT_TRUE(result) << "Failed to retrieve current state from Lifecycle.state() method";
    EXPECT_EQ(*result, Firebolt::Lifecycle::LifecycleState::INITIALIZING);
}

TEST_F(LifecycleTest, stateBadResponse)
{
    mock_with_response("Lifecycle2.state", "invalid_response");
    Firebolt::Result<Firebolt::Lifecycle::LifecycleState> result = lifecycleImpl_.state();
    ASSERT_FALSE(result) << "LifecycleImpl::state() did not return an error";
}

TEST_F(LifecycleTest, subscribeOnStateChanged)
{
    EXPECT_CALL(mockHelper, subscribe(_, "Lifecycle2.onStateChanged", _, _))
        .WillOnce(Invoke([&](void* /*owner*/, const std::string& /*eventName*/, std::any&& /*notification*/,
                             void (* /*callback*/)(void*, const nlohmann::json&))
                         { return Firebolt::Result<Firebolt::SubscriptionId>{1}; }));

    EXPECT_CALL(mockHelper, unsubscribe(1)).WillOnce(Return(Firebolt::Result<void>{Firebolt::Error::None}));

    auto id =
        lifecycleImpl_.subscribeOnStateChanged([](const std::vector<Firebolt::Lifecycle::StateChange>& /*changes*/) {});

    ASSERT_TRUE(id) << "error on subscribe ";
    EXPECT_EQ(id.value(), 1);

    auto result = lifecycleImpl_.unsubscribe(id.value());
    ASSERT_TRUE(result) << "error on unsubscribe";
}
